lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -216,33 +218,38 @@ mkDBFactory
        -- ^ Logging object
     -> Maybe FilePath
        -- ^ Path to database directory, or Nothing for in-memory database
-    -> DBFactory IO s k
-mkDBFactory cfg tr mDatabaseDir = case mDatabaseDir of
-    Nothing -> DBFactory
-        { withDatabase = \_ ->
-            withDBLayer cfg tracerDB Nothing
-        , removeDatabase = \_ ->
-            pure ()
-        }
-    Just databaseDir -> DBFactory
-        { withDatabase = \wid ->
-            withDBLayer cfg tracerDB (Just $ databaseFile wid)
-        , removeDatabase = \wid -> do
-            let files =
-                    [ databaseFile wid
-                    , databaseFile wid <> "-wal"
-                    , databaseFile wid <> "-shm"
-                    ]
-            mapM_ removePathForcibly files
-        }
+    -> IO (DBFactory IO s k)
+newDBFactory cfg tr mDatabaseDir = do
+    mvar <- newEmptyMVar
+    case mDatabaseDir of
+        Nothing -> pure DBFactory
+            { withDatabase = \_ action ->
+                withDBLayer cfg tracerDB Nothing (action . snd)
+            , removeDatabase = \_ ->
+                pure ()
+            }
+        Just databaseDir -> pure DBFactory
+            { withDatabase = \wid action ->
+                withDBLayer cfg tracerDB (Just $ databaseFile wid) $ \(ctx,db) -> do
+                    putMVar mvar ctx
+                    action db
+            , removeDatabase = \wid -> do
+                let files =
+                        [ databaseFile wid
+                        , databaseFile wid <> "-wal"
+                        , databaseFile wid <> "-shm"
+                        ]
+                readMVar mvar >>= destroyDBLayer
```

Comment:
That's the main change here. Make sure the db is closed before we attempt to remove the files from the disk ^.^ ...